input.cpp

```diff
@@ -2717,12 +2732,18 @@ int input_test(int getchar)
 	if (state == 1)
 	{
 		printf("Open up to %d input devices.\n", NUMDEV);
+
+		// Remember which devices mapped to which core device numbers.
+		std::map<std::string, unsigned int> devnums;
```

Comment:
Actually, there's one more problem -- device numbers aren't globally/permanently remembered during the lifetime of the MiSTer process, so as different devices get disconnected/reconnected, the mapping of real device to virtual device changes around... and sometimes the keyboard ends up mapped to a `num` that was previously used for gamepad1, so then neither of the gamepads work.  I think the right thing to do is to make this a global/static, and then move all the logic into that same loop I just rewrote, so all the logic for assigning input device `num`s is all in one place.

Reply:
I've committed another change to this pending PR which implements that idea.  Now, no matter what order I disconnect/reconnect devices, the same device always maps to the same virtual device, as the user would naturally expect.

Reply:
It's already fixed in release. If assigned device is removed then it's removed from player assignment forever.